### PR TITLE
MULE-13780: Optimize streaming handler.

### PR DESCRIPTION
**Pull request by afelisatti**
DO NOT MERGE: Review only for now.

### CODE REVIEW

File: src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java [hunk 1]
```diff
@@ -103,6 +100,14 @@ public STATE onHeadersReceived(HttpResponseHeaders headers) throws Exception {
 
   @Override
   public STATE onBodyPartReceived(HttpResponseBodyPart bodyPart) throws Exception {
+    if (input == null && bodyPart.isLast()) {
```
Comment (elrodro83) [inline src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java]: ask for `input == null` only once, for readability

File: src/test/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandlerTestCase.java [hunk 2]
```diff
@@ -34,16 +43,65 @@
...
+    Reference<InputStream> responseContent = new Reference<>();
+    ResponseBodyDeferringAsyncHandler handler = new ResponseBodyDeferringAsyncHandler(future, 1024);
+    handler.onStatusReceived(mock(HttpResponseStatus.class, RETURNS_DEEP_STUBS));
+    GrizzlyResponseBodyPart bodyPart = mock(GrizzlyResponseBodyPart.class, RETURNS_DEEP_STUBS);
+    when(bodyPart.isLast()).thenReturn(false);
+    future.whenComplete((response, exception) -> responseContent.set(response.getEntity().getContent()));
+
+    assertThat(handler.onBodyPartReceived(bodyPart), is(CONTINUE));
+
+    new PollingProber(5000, 300).check(new JUnitProbe() {
```
Comment (elrodro83) [inline src/test/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandlerTestCase.java]: use constants

File: src/test/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandlerTestCase.java [hunk 3]
```diff
@@ -34,16 +43,65 @@
...
 public class ResponseBodyDeferringAsyncHandlerTestCase extends AbstractMuleTestCase {
 
+  private static final int PROBE_TIMEOUT = 5000;
+  private static final int POLL_DELAY = 300;
+
+  @Test
+  public void doesNotStreamWhenPossible() throws Exception {
+    CompletableFuture<HttpResponse> future = new CompletableFuture<>();
+    Reference<InputStream> responseContent = new Reference<>();
+    ResponseBodyDeferringAsyncHandler handler = new ResponseBodyDeferringAsyncHandler(future, 1024);
```
Comment (elrodro83) [inline src/test/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandlerTestCase.java]: extract the value to a  constant

File: src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java [hunk 4]
```diff
@@ -47,18 +45,17 @@
   private static final Logger logger = LoggerFactory.getLogger(ResponseBodyDeferringAsyncHandler.class);
 
   private volatile Response response;
-  private final OutputStream output;
-  private final InputStream input;
+  private final int bufferSize;
+  private PipedOutputStream output;
+  private InputStream input;
```
Comment (elrodro83) [inline src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java]: make this an Optional?

File: src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java [hunk 5]
```diff
@@ -46,16 +53,28 @@
 public class ResponseBodyDeferringAsyncHandler implements AsyncHandler<Response> {
 
   private static final Logger logger = LoggerFactory.getLogger(ResponseBodyDeferringAsyncHandler.class);
+  private static final int SIZE_17_KB = 17 * 1024;
```
Comment (elrodro83) [inline src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java]: add a comment explaining how you get to this number

File: src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java [hunk 6]
```diff
@@ -46,16 +53,28 @@
...
   private final Response.ResponseBuilder responseBuilder = new Response.ResponseBuilder();
   private final HttpResponseCreator httpResponseCreator = new HttpResponseCreator();
   private final AtomicBoolean handled = new AtomicBoolean(false);
 
+  static {
+    try {
+      responseField = HttpResponseHeaders.class.getDeclaredField("response");
+      responseField.setAccessible(true);
+    } catch (NoSuchFieldException e) {
+      // eat it
```
Comment (elrodro83) [inline src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java]: a log at least?

File: src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java [hunk 7]
```diff
@@ -98,12 +117,36 @@ public STATE onStatusReceived(HttpResponseStatus responseStatus) throws Exceptio
   @Override
   public STATE onHeadersReceived(HttpResponseHeaders headers) throws Exception {
     responseBuilder.accumulate(headers);
+    calculateBufferSize(headers);
     return CONTINUE;
   }
 
+  private void calculateBufferSize(HttpResponseHeaders headers) throws IllegalAccessException {
+    String contentLength = headers.getHeaders().get(CONTENT_LENGTH).get(0);
```
Comment (afelisatti) [inline src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java]: Wouldn't this throw an NPE if there's none?

File: src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java (hunk 5 above)
Comment (afelisatti) [reply src/main/java/org/mule/service/http/impl/service/client/async/ResponseBodyDeferringAsyncHandler.java]: I think we should use the user's in this case and put the default at the attribute level.

**Pull request merged**